File: 2020/01/part2.py
#!/usr/bin/env python3

pairs = {}
numbers = [int(line) for line in open('input.txt', 'r', encoding='ascii')]

for n in numbers:
    for k in numbers:
        pairs[n + k] = n * k

for n in numbers:
    m = 2020 - n
    if m in pairs:
        print(n * pairs[m])

File: 2021/01/part2.hs
#!/usr/bin/env runhaskell

sum3 :: [Int] -> [Int]
sum3 xs = id
  -- Group the elements of lists with the first 0..2 elements dropped,
  -- then sum the groups.
  $ fmap sum
  $ foldr (\n -> zipWith (:) (drop n xs)) (repeat []) [0..2]

solve :: [Int] -> Int
solve xs =
  let
    xs3 = sum3 xs
  in
    length $ filter (\(x, y) -> y > x) (zip xs3 $ tail xs3) 

main :: IO ()
main = readFile "input.txt" >>= print . solve . fmap read . lines
